staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -492,21 +492,22 @@ func (s sortableStoreElements) Swap(i, j int) {
 
 // WaitUntilFreshAndList returns list of pointers to `storeElement` objects along
 // with their ResourceVersion and the name of the index, if any, that was used.
-func (w *watchCache) WaitUntilFreshAndList(ctx context.Context, resourceVersion uint64, matchValues []storage.MatchValue) ([]interface{}, uint64, string, error) {
-	var err error
+func (w *watchCache) WaitUntilFreshAndList(ctx context.Context, resourceVersion uint64, matchValues []storage.MatchValue) (result []interface{}, rv uint64, index string, err error) {
 	if utilfeature.DefaultFeatureGate.Enabled(features.ConsistentListFromCache) && w.notFresh(resourceVersion) {
 		w.waitingUntilFresh.Add()
 		err = w.waitUntilFreshAndBlock(ctx, resourceVersion)
 		w.waitingUntilFresh.Remove()
 	} else {
 		err = w.waitUntilFreshAndBlock(ctx, resourceVersion)
 	}
+
+	defer func() { sort.Sort(sortableStoreElements(result)) }()
```

Comment:
Readability: Using defer to sort data creates quite non-obvious code flow.

Maybe moving w.waitUnitilFreshAndBlock calls to the anonymous function in line 510 and defer w.RUnlock() there + keeping sort where it was could simplify the flow here?

Reply:
That was what I did initially, but sorting in defer is more future-proof imho as it ensures that we will not skip it in potential new code paths in the future.